test/Microsoft.ML.TestFramework/BaseTestBaseline.cs

```diff
@@ -75,6 +75,7 @@ protected BaseTestBaseline(ITestOutputHelper output) : base(output)
         private string _baselineCommonDir;
         private string _baselineBuildStringDir;
         private IEnumerable<string> _baselineConfigDirs;
+        private HashSet<string> _usedSpecificBaselineConfigs;
 
```

Comment:
Why do you need a HashSet?

Reply:
Because more than one specific configuration can be used. For example, using a netcoreapp3.1 and linux-x64 baseline.

Reply:
Sorry I am not following. Can you please walk me through the scenario end to end when the tests are running where you would see the length of `_usedSpecificBaselineConfigs' being greater than 1?

---
In reply to: [415395501](https://github.com/dotnet/machinelearning/pull/5060#discussion_r415395501) [](ancestors = 415395501)

Reply:
Sure. For example, for `BinaryPriorTest`, if a baseline is available at, say `test\BaselineOutput\Common\PriorPredictor\linux-x64\netcoreapp31\BinaryPrior-TrainTest-breast-cancer-out.txt`, there would be 2 specific configurations for which a baseline is available: `linux-x64` and `netcoreapp31`. In this case, it would be correct to say that the baseline being used is for Linux x64 **and** NET Core 3.1, and hence `_usedSpecificBaselineConfigs` would have length of 2.

Of course, this assumes that there may be baselines available for a test that differs on Linux x64 and NET Core 3.1 together, that are different than the standard baselines. Currently, no such baselines exist. And the current implementation here does not support the case where a baseline would be in the folder `\linux-x64\netcoreapp31\` or `\linux-x64-netcoreapp31\`.

https://github.com/dotnet/machinelearning/blob/bfd2773dd59ac302bf659ed6959738aa68dd5756/test/Microsoft.ML.TestFramework/BaseTestBaseline.cs#L250-L279

https://github.com/dotnet/machinelearning/blob/bfd2773dd59ac302bf659ed6959738aa68dd5756/test/Microsoft.ML.TestFramework/BaseTestBaseline.cs#L114-L151

However, I believe this scenario is possible and can occur in the future, hence this PR's implementation. Of course, it doesn't make sense to use a HashSet if we assume that no two specific configurations can occur at the same time, or that if it does, it will be in a folder named, say `\linux-x64-netcoreapp31\`. But we should decide on how to handle baselines in this case, so that I can keep or change this PR's implementation.

Reply:
> > Test BinaryPriorTest is using linux-x64, netcoreapp31 configuration specific baselines.

Let us assume that different baselines existed for a particular test. Under what conditions will you see this statement in the log? Can you please attempt a repro by simply copying the current baselines into a new directory of the desired configuration?

Reply:
As I said above, in the current implementation of the code I posted above, you won't ever see this exact statement in the log. This is due to the way in which these configuration-specific paths are checked: it only checks 1 possible configuration at a time.

As you asked, I copied the baseline lines into the directory `test\BaselineOutput\Common\PriorPredictor\win-x64\netcoreapp31\`, but, again, this path was never looked at. The only paths that were checked in `BaseTestBaseline.cs` were:

- test\BaselineOutput\Common\PriorPredictor\win-x64\
- test\BaselineOutput\Common\PriorPredictor\netcoreapp31\
This behavior is expected because again, only 1 specific configuration possibility is being checked at a time.

This is why I said above we should decide whether or not there can be baselines specific to 2 or more specific-configurations (such as a linux-x64-netcoreapp31 baseline). If not, there's no need of a HashSet and a simple string can be used to store a used specific configuration for logging. If there can be 2 or more specific configurations that result in brand-new baselines, then changes to `GetBaselinePath(string subDir, string name)` is necessary to reflect the possibility of there being differing baselines when 2 or more specific configurations are used. @harishsk @frank-dong-ms Do you expect a situation in which 2 specific baselines can result in brand-new baselines?